pippy/PipelineSchedule.py

```diff
@@ -74,22 +74,225 @@ def compute_loss(self):
 ) -> List[torch.Tensor]:
     """
-    Creates buffers for a given input on a specified device.
+    Creates buffers for a given tensor on a specified device.
     This function takes as input a tensor or a list of tensors and returns a tensor or a list of tensors (respectively)
     of the same shape, but located on the specified device and uninitialized (i.e., filled with arbitrary data).
     """
-    if isinstance(input, torch.Tensor):
-        return [torch.empty_like(input, device=device)]
-    elif isinstance(input, (list, tuple)):
-        return [torch.empty_like(inp, device=device) for inp in input]
-    raise ValueError(
-        f"Unsupported input type {type(input)} cannot create buffers"
+    if isinstance(tensor, torch.Tensor):
+        return [torch.empty_like(tensor, device=device)]
+    elif isinstance(tensor, (list, tuple)):
+        return [torch.empty_like(t, device=device) for t in tensor]
+    raise TypeError(
+        f"Unsupported input type {type(tensor)} cannot create buffers"
     )
 
 
+METADATA_TENSOR_LEN = 50
+PLACEHOLDER_VAL = 0
+
+
+def create_metadata_tensor(
+    tensors: Optional[List[torch.Tensor]] = None,
+    device: Optional[torch.device] = torch.device("cpu"),
+) -> torch.Tensor:
+    """
+    Create a metadata tensor that can be sent over the wire.
+    This tensor contains the number of dimensions and the shape of each tensor being sent.
+    If the tensor is None, then this tensor will contain 0s.
+    """
+    metadata_tensor = torch.full(
+        (METADATA_TENSOR_LEN,),
+        PLACEHOLDER_VAL,
+        dtype=torch.int32,
+        device=device,
+    )
+    if tensors:
+        # Create a list of tensors containing the number of dimensions and the shape of each tensor
+        data = [
+            # data is of format [num_dims, dim1, dim2, ...]
+            torch.tensor(
+                [len(tensor.shape)] + list(tensor.shape),
+                dtype=torch.int32,
+                device=device,
+            )
+            for tensor in tensors
+        ]
+        # Concatenate the data into a single tensor
+        data_tensor = torch.cat(data)
+        dt_shape = data_tensor.shape[0]
+        if dt_shape > METADATA_TENSOR_LEN:
+            raise ValueError(
+                f"Metadata tensor size ({dt_shape}) exceeds maximum allowed length ({METADATA_TENSOR_LEN})."
+            )
+        metadata_tensor[:dt_shape] = data_tensor
+    return metadata_tensor
+
+
+def extract_metadata_from_tensor(tensor: torch.Tensor) -> List[List[int]]:
+    """
+    Extract the number of dimensions and the shape of each tensor from a metadata tensor.
+    """
+    metadata = []
+    i = 0
+    while i < len(tensor) and tensor[i] != PLACEHOLDER_VAL:
+        num_dims = tensor[i].item()
+        shape = tensor[i + 1 : i + 1 + num_dims].tolist()
+        metadata.append(shape)
+        i += num_dims + 1
+    return metadata
+
+
+def dry_run(
+    models: List[nn.Module],
```

Comment:
nit: does `models: List[nn.Module]` here refer to a full model or the nn.Module's assigned to this rank?

Reply:
nn.Modules assigned to this rank, so the length should be 1 for any non-interleaved schedules and >1 for any interleaved schedules